From d109e64eb37626ffb97320f0ab98788245fbcc51 Mon Sep 17 00:00:00 2001
From: thippesw <thippesw@xilinx.com>
Date: Thu, 9 Sep 2021 03:41:22 -0600
Subject: [PATCH] return val changes

---
 drivers/misc/xilinx_pci_endpoint.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/misc/xilinx_pci_endpoint.c b/drivers/misc/xilinx_pci_endpoint.c
index 7fb0dcd97aef..a8ebb963d29d 100644
--- a/drivers/misc/xilinx_pci_endpoint.c
+++ b/drivers/misc/xilinx_pci_endpoint.c
@@ -711,10 +711,8 @@ static int pciep_dmabuf_alloc(struct pciep_driver_data *this, char __user *argp)
 	dev_err(this->dma_dev, "%s copy to user failed\n",__func__);
 	goto error_buf_fd;
     }
-    return 0;
-
+    goto success;
 err_buf_fd :
-
     dma_buf_put(dmabuf);
 err1 :
     for (i = 0; i < NUM_BUFFERS; i++) {
@@ -726,6 +724,8 @@ err1 :
     bp.size = 0;
 err_alloc :
     return -1;
+success :
+	return 0;
 
 }
 
-- 
2.17.1

